### PR TITLE
feat(protocol-designer): tip positioning modal and assets

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-928

# Overview

update tip position modals for all advanced settings

Screenshot 2024-10-28 at 08 42 31

Screenshot 2024-10-25 at 16 17 20

## Test Plan and Hands on Testing

Test each tip positioning field and test that all functionality of the modals work

Here is how to access the modal:
1. aspirate tip position field (x,y,z)
2. aspirate delay advanced settings tip position field (z only)
3. aspirate touch tip advanced settings tip position field (z only)  - this one should start from the top of the well instead of the bottom
4. dispense tip position field (x,y,z)
5. blowout advanced settings tip position field (z only)
6. blowout touch tip advanced settings tip position field (z only) - this one should start from the top of the well instead of the bottom

## Changelog

- add svgs
- create visualization components for side view, top view, and z-position only side view
- create new modals for tip positioning for z only and x,y,z positions
- wire in the modals

## Risk assessment

low

### CODE REVIEW

**Comment by jerader** [comment]
todo:

- [x] solve merge conflicts
- [x] update copy to match designs

File: protocol-designer/src/assets/localization/en/shared.json [hunk 1]
```diff
@@ -125,6 +126,14 @@
...
   "step": "Step {{current}} / {{max}}",
   "consent_to_eula": "By using Protocol Designer, you consent to the Opentrons <link1>EULA</link1>.",
+  "swap_view": "Swap view",
+  "tip_position": "Edit {{prefix}} tip position",
+  "tip_position_blowout_z_offset": "Edit blowout position",
+  "tip_position_aspirate_touchTip_mmFromBottom": "Edit aspirate touch tip position",
+  "tip_position_dispense_touchTip_mmFromBottom": "Edit dispense touch tip position",
+  "tip_position_mix_touchTip_mmFromBottom": "Edit touch tip position",
+  "tip_position_aspirate_delay_mmFromBottom": "Edit aspirate delay position",
+  "tip_position_dispense_delay_mmFromBottom": "Edit dispense delay position",
```
Comment (ncdiehl11) [inline protocol-designer/src/assets/localization/en/shared.json]:
```suggestion
Suggested change
  "tip_position": "Edit {{prefix}} tip position",
  "tip_position": "Edit {{prefix}} tip position",
  "tip_position_aspirate_delay_mmFromBottom": "Edit aspirate delay position",
  "tip_position_aspirate_touchTip_mmFromBottom": "Edit aspirate touch tip position",
  "tip_position_blowout_z_offset": "Edit blowout position",
  "tip_position_dispense_delay_mmFromBottom": "Edit dispense delay position",
  "tip_position_dispense_touchTip_mmFromBottom": "Edit dispense touch tip position",
  "tip_position_mix_touchTip_mmFromBottom": "Edit touch tip position",
```

File: protocol-designer/src/organisms/TipPositionModal/TipPositionSideView.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    : mmFromBottom - WELL_HEIGHT_PIXELS
+
+  const xPx = (WELL_WIDTH_PIXELS / xWidthMm) * xPosition
+  const roundedXPx = round(xPx, PIXEL_DECIMALS)
+
+  return (
+    <Box
+      position={POSITION_RELATIVE}
+      width="253px"
+      height="288px"
```
Comment (ncdiehl11) [inline protocol-designer/src/organisms/TipPositionModal/TipPositionSideView.tsx]: let's use rem

File: protocol-designer/src/organisms/TipPositionModal/TipPositionTopView.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  const yPx = (WELL_WIDTH_PIXELS / yWidthMm) * yPosition
+
+  const roundedXPx = round(xPx, PIXEL_DECIMALS)
+  const roundedYPx = round(yPx, PIXEL_DECIMALS)
+  const translateY = roundedYPx < 0 ? Math.abs(roundedYPx) : -roundedYPx
+  return (
+    <Box
+      position={POSITION_RELATIVE}
+      width="253px"
+      height="288px"
```
Comment (ncdiehl11) [inline protocol-designer/src/organisms/TipPositionModal/TipPositionTopView.tsx]: same

File: protocol-designer/src/organisms/TipPositionModal/TipPositionZOnlyView.tsx
Comment (ncdiehl11) [inline protocol-designer/src/organisms/TipPositionModal/TipPositionZOnlyView.tsx:1]: I wonder if we should refactor this into `TipPositionSideView`? If you feel the content diverges significantly enough to be in its own component, I understand that as well.

File: protocol-designer/src/organisms/TipPositionModal/utils.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  const errors: Error[] = []
+
+  const value = Number(rawValue)
+  if (rawValue === null || Number.isNaN(value)) {
+    // blank or otherwise invalid should show this error as a fallback
+    return [OUT_OF_BOUNDS]
+  }
+  const incorrectDecimals = round(value, DECIMALS_ALLOWED) !== value
+  const outOfBounds = value > maxMm || value < minMm
+  console.log(outOfBounds)
```
Comment (ncdiehl11) [inline protocol-designer/src/organisms/TipPositionModal/utils.tsx]: can remove

File: protocol-designer/src/organisms/TipPositionModal/utils.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  const { value: rawValue, maxMm, minMm } = args
+  const errors: Error[] = []
+
+  const value = Number(rawValue)
+  if (rawValue === null || Number.isNaN(value)) {
+    // blank or otherwise invalid should show this error as a fallback
+    return [OUT_OF_BOUNDS]
+  }
+  const incorrectDecimals = round(value, DECIMALS_ALLOWED) !== value
+  const outOfBounds = value > maxMm || value < minMm
```
Comment (ncdiehl11) [inline protocol-designer/src/organisms/TipPositionModal/utils.tsx]: nit, maybe use `hasIncorrectDecimals`/`isOutOfBounds`

File: protocol-designer/src/organisms/TipPositionModal/ZTipPositionModal.tsx [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    errors,
+    minMm,
+    maxMm,
+    isPristine,
+    t,
+  })
+
+  const handleDone = (): void => {
+    if (!hasErrors) {
+      updateValue(value === null ? null : Number(value))
```
Comment (ncdiehl11) [inline protocol-designer/src/organisms/TipPositionModal/ZTipPositionModal.tsx]: `== null`

Comment (ncdiehl11) [reply protocol-designer/src/organisms/TipPositionModal/ZTipPositionModal.tsx]: Also, looks like you can save even with empty values. Should we update this?

**Review by ncdiehl11 — COMMENTED** [review]
mostly looks great but left some comments that we should address before merging

File: protocol-designer/src/organisms/TipPositionModal/TipPositionZOnlyView.tsx
Comment (jerader) [reply protocol-designer/src/organisms/TipPositionModal/TipPositionZOnlyView.tsx:1]: good call out, i thought about merging them but found it easier to keep them separate, i also didn't try hard though lol. i'll look back into this.

File: protocol-designer/src/organisms/TipPositionModal/utils.tsx (hunk 5 above)
Comment (jerader) [reply protocol-designer/src/organisms/TipPositionModal/utils.tsx]: thanks, this was copied from legacy code, definitely didn't try to refine it 😆

File: protocol-designer/src/organisms/TipPositionModal/ZTipPositionModal.tsx (hunk 6 above)
Comment (jerader) [reply protocol-designer/src/organisms/TipPositionModal/ZTipPositionModal.tsx]: i think when its empty, it defaults to 0mm, so i don't think its blocking but let me check it out still

File: protocol-designer/src/organisms/TipPositionModal/TipPositionZOnlyView.tsx
Comment (jerader) [reply protocol-designer/src/organisms/TipPositionModal/TipPositionZOnlyView.tsx:1]: i think i'm going to keep them separate, there is a lot of small differences between the 2

File: protocol-designer/src/assets/images/tip_side_bottom_layer.svg
Comment (koji) [inline protocol-designer/src/assets/images/tip_side_bottom_layer.svg:1]: I cannot see anything. Is this expected?

File: protocol-designer/src/organisms/TipPositionModal/TipPositionSideView.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  const { t } = useTranslation('application')
+  const fractionOfWellHeight = mmFromBottom / wellDepthMm
+  const pixelsFromBottom =
+    fractionOfWellHeight * WELL_HEIGHT_PIXELS - WELL_HEIGHT_PIXELS
+  const roundedPixelsFromBottom = round(pixelsFromBottom, PIXEL_DECIMALS)
+  const bottomPx = wellDepthMm
+    ? roundedPixelsFromBottom * 2
+    : mmFromBottom - WELL_HEIGHT_PIXELS
+
+  const xPx = (WELL_WIDTH_PIXELS / xWidthMm) * xPosition
```
Comment (koji) [inline protocol-designer/src/organisms/TipPositionModal/TipPositionSideView.tsx]: maybe something like `pixelsPerMm` ?

File: protocol-designer/src/assets/images/tip_side_bottom_layer.svg
Comment (jerader) [reply protocol-designer/src/assets/images/tip_side_bottom_layer.svg:1]: um i'm not sure but it works correctly when rendering it in the component?

**Review by ncdiehl11 — APPROVED** [review]
nice 🚀

**Pull request merged**